### DISCUSSION TITLE
STM32WL55 on a E77-400M22S how do I get REPL?

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13110] [STM32 / Pyboard] [answered]
I seem to have been able to load the STM32WL55 hex firmware
on to this device.  I used STM32 CubeProgrammer and a ST-Link.

As there is no USB I thought ... how am I going to get REPL
going so that I can work on the actual device?

In the mpconfigboard.h for the STM32WL55 it states:
```
// UART buses
#define MICROPY_HW_UART1_TX                     (pin_B6)    // Arduino D1, pin 7 on CN9
#define MICROPY_HW_UART1_RX                     (pin_B7)    // Arduino D0, pin 8 on CN9
#define MICROPY_HW_LPUART1_TX                   (pin_A2)    // ST-link
#define MICROPY_HW_LPUART1_RX                   (pin_A3)    // ST-link
#define MICROPY_HW_UART_REPL                    PYB_LPUART_1
#define MICROPY_HW_UART_REPL_BAUD               115200
```
Should I be using A2 and A3?  Also don't  I need to use a FTDI programmer?  What is meant by the ST-Link comment?  

Appreciate any suggestions on how to get to the next stage.

**Comment by andrewleech** [comment, marked as answer]
Yes the NUCLEO_WL55 board / firmware is configured with repl on LPUART1 on pins PA2 and PA3

The nucleo board comes with a built in programmer & USB com port interface called stlink, that's what the comment is referring to: https://www.st.com/en/evaluation-tools/nucleo-wl55jc.html

Any USB UART interface like ftdi should work though.

**Reply by davefes** [reply]
Thanks for the reply.  Tried A2 and A3 and didn't see anything ...maybe the firmware didn't make its way to the flash.  STM32CubeProgrammer didn't complain.

I'll try finding a schematic for the Nucleo-WL55 board and see if I need to do anything special in setting up the device.

**Reply by andrewleech** [reply]
Does your board have a 32khz crystal? I think the nucleos often do, a mismatch in that setting is a common reason for failed startups.

**Reply by davefes** [reply]
> The external crystal oscillator uses 32.768KHz, 32MHz high-precision industrial-grade crystal oscillator

I saw a picture of the insides with a 32.768KHz shown on it, as well as the 32MHz unit.

I have contacted support at cdebyte.com  In their rather skant datasheet they show using J-Link to program the part.

**Reply by andrewleech** [reply]
Taking another look at the nucleo settings: https://github.com/micropython/micropython/blob/b5449b0f095a2232f38ca7842614dcfda83ffc26/ports/stm32/boards/NUCLEO_WL55/mpconfigboard.h#L37

Bypass mode on the clock is enabled because on that board it doesn't use an external 32mhz crystal, but instead gets a logic level clock signal from the stlink section.

**Reply by davefes** [reply]
Thanks for looking into this for me.  Is the way forward (maybe) for me to build the firmware changing that to:
`#define MICROPY_HW_CLK_USE_BYPASS               (0) `
?

**Reply by andrewleech** [reply]
Yep that would be the next step certainly.

**Reply by davefes** [reply]
In process ... thank you.

**Reply by davefes** [reply]
Same result.
```
dave@davef:~$ rsusb
Using buffer-size of 30
Connecting to /dev/ttyUSB0 (buffer-size 30)...
```
after a minute or so if I hit CTRL-C I get a screen-full of messages but I think those are only because something times-out in rshell
mpremote is non-responsive.

Edit:  NRST does nothing as well.

**Reply by davefes** [reply]
> Note 1: PA6 and PA7 pins are used as internal control RF switches of the module, PA6 = RF_TXEN, PA7 = RF_RXEN,
RF_TXEN=1 RF_RXEN=0 is the transmit channel, and RF_TXEN=0 RF_RXEN=1 is the receiving channel
Note 2: The PC14-OSC32_IN and PC15-OSC32_OUT pins have been connected to a 32.768KHz crystal oscillator inside
the module for users to choose and use in secondary development.
Note 3: The OSC_IN and OSC_OUT pins have been connected to a 32MHz crystal oscillator inside the module for
users to choose and use in secondary development.

I was going to make this a TX board so I'll try PA6 = RF_TXEN, PA7 = RF_RXEN,
RF_TXEN=1 RF_RXEN=0 for transmit

Ah, PA6 and PA7 do not come to the module interface.  Looks like they have be programmatically set.

**Comment by davefes** [comment]
Check your pin assignments very carefully!!

**Comment by projectgus** [comment]
Note that the microcontroller in E77-400M22S is STM32WLE5, which is different to the NUCLEO_WL55 board which has STM32WL55.

I think the two microcontrollers are very similar, but they are likely not 100% the same and other firmware changes might be needed to bring MicroPython up on the other board.

**Reply by davefes** [reply]
I have Micropython running on the board with LFS2, but quickly ran out of flash.  I couldn't load stm32wl5.py, but will try making a .mpy file out of it.

**Comment by davefes** [comment]
Must have accidentally closed this ... could be useful information.

**Reply by maholli** [reply]
Appreciate you keeping this discussion around- some of us are still stumbling onto it!

Would you mind listing the changes you made to the `NUCLEO_WL55` build that allowed it to run on the `STM32WLE5` chips? Any amount would help, as I'm finding much of the referenced "stm32wl55" build files should work for "wle5" as well.

**Reply by davefes** [reply]
Have you picked-up on the other related-conversations?  My stumbling-blocks were:
1) you can load only small files, ie 1-2KB 
2) could NOT find what was causing `RuntimeError: Internal radio Status (2, 1) OpError 0x20`  The manufacturer could not help.

I think it has to do with the way the internal oscillators are set so I built 3 versions of firmware to try to see what was missing.  I can supply those if wanted.

My changes to `mpconfigboard.h`:
```py
/* This file is part of the MicroPython project, https://micropython.org/
 * MIT License; Copyright (c) 2021 Damien P. George
 */

#define MICROPY_HW_BOARD_NAME                   "NUCLEO-WL55"
#define MICROPY_HW_MCU_NAME                     "STM32WL55JCI7"

#define MICROPY_EMIT_THUMB                      (0)
#define MICROPY_EMIT_INLINE_THUMB               (0)
#define MICROPY_PY_BUILTINS_COMPLEX             (0)
#define MICROPY_PY_GENERATOR_PEND_THROW         (0)
#define MICROPY_PY_MATH                         (0)
#define MICROPY_PY_FRAMEBUF                     (0)
#define MICROPY_PY_SOCKET                       (0)
#define MICROPY_PY_NETWORK                      (0)
#define MICROPY_PY_ONEWIRE                      (0)
#define MICROPY_PY_STM                          (1) // for subghz radio functions
#define MICROPY_PY_STM_CONST                    (0) // saves size, no named registers
#define MICROPY_PY_PYB_LEGACY                   (0)
#define MICROPY_PY_HEAPQ                        (0)

#define MICROPY_HW_HAS_FLASH                    (1)
#define MICROPY_HW_ENABLE_RTC                   (1)
#define MICROPY_HW_ENABLE_RNG                   (1)
#define MICROPY_HW_ENABLE_ADC                   (0) // use machine.ADC instead
#define MICROPY_HW_HAS_SWITCH                   (1)

// There is an external 32kHz oscillator
#define MICROPY_HW_RTC_USE_LSE                  (1)
#define MICROPY_HW_RTC_USE_US                   (1)

// Use external 32MHz TCXO + PLL as system clock source
// (If unset, board will use the internal MSI oscillator instead.)
#define MICROPY_HW_CLK_USE_HSE                  (0)  // for the E77-400M22S

// HSE bypass for STM32WL5x means TCXO is powered from PB0_VDDTCXO pin
#define MICROPY_HW_CLK_USE_BYPASS               (1)

// UART buses
#define MICROPY_HW_UART1_TX                     (pin_B6)    // Arduino D1, pin 7 on CN9
#define MICROPY_HW_UART1_RX                     (pin_B7)    // Arduino D0, pin 8 on CN9
#define MICROPY_HW_LPUART1_TX                   (pin_A2)    // ST-link
#define MICROPY_HW_LPUART1_RX                   (pin_A3)    // ST-link
#define MICROPY_HW_UART_REPL                    PYB_LPUART_1
#define MICROPY_HW_UART_REPL_BAUD               115200

// I2C buses
#define MICROPY_HW_I2C1_SCL                     (pin_A12)   // Arduino D15, pin 10 on CN5
#define MICROPY_HW_I2C1_SDA                     (pin_A11)   // Arduino D14, pin 9 on CN5
#define MICROPY_HW_I2C3_SCL                     (pin_B13)   // Arduino A5, pin 6 on CN8
#define MICROPY_HW_I2C3_SDA                     (pin_B14)   // Arduino A4, pin 5 on CN8

// SPI buses, do NOT use this bus on the E77-400M22S
//#define MICROPY_HW_SPI1_NSS                     (pin_A4)    // Arduino D10  pin 3 on CN5
//#define MICROPY_HW_SPI1_SCK                     (pin_A5)    // Arduino D13, pin 6 on CN5
//#define MICROPY_HW_SPI1_MISO                    (pin_A6)    // Arduino D12, pin 5 on CN5
//#define MICROPY_HW_SPI1_MOSI                    (pin_A7)    // Arduino D11, pin 4 on CN5

// SUBGHZSPI Internal radio SPI BUS
#define MICROPY_HW_SUBGHZSPI_NAME               "SUBGHZ"
#define MICROPY_HW_SUBGHZSPI_ID                 3

// User switch; pressing the button makes the input go low
#define MICROPY_HW_USRSW_PIN                    (pin_A0)
#define MICROPY_HW_USRSW_PULL                   (GPIO_PULLUP)
#define MICROPY_HW_USRSW_EXTI_MODE              (GPIO_MODE_IT_FALLING)
#define MICROPY_HW_USRSW_PRESSED                (0)

// LEDs
#define MICROPY_HW_LED1                         (pin_B15) // blue
#define MICROPY_HW_LED2                         (pin_B9) // green
#define MICROPY_HW_LED3                         (pin_B11) // red
#define MICROPY_HW_LED_ON(pin)                  (mp_hal_pin_high(pin))
#define MICROPY_HW_LED_OFF(pin)                 (mp_hal_pin_low(pin))
```
With regard to 1) because my code is so fluid I am reluctant to freeze stuff ... because I feel like I am going back to ATMega88 stuff.
Here is a snippet from other discussions:
```
The STM32WL55 is a pretty low-resource device, so it doesn't have a flash filesystem automatically. You don't always need a filesystem in the internal flash, for example if you freeze all of the Python code into the firmware binary and flash it that way.

There is support in the default WL55 config for Littlefs2, so you can format a Littlefs image in the available flash space like this:

import os, pyb
os.VfsLfs2.mkfs(pyb.Flash(start=0))

... then when you next restart (for example, with Ctrl-D) this filesystem will be mounted at /flash and no more errors. The documentation for this feature is here.

Note the available flash will be small, on my system it was 24KB. Run pyb.Flash(start=0) to print the value of the Flash object, which includes the length.
```
Good luck!  If you make progress I would sure like to know as I think this device would make a nice LoRa link.

**Reply by maholli** [reply]
Thank you for summarizing where you landed on this! I agree about the potential these little SoC have. I'll follow up once I have a chance to work with some hardware.

For what it's worth, I'm coming at this from the RAK3172 perspective as opposed to your E77-400M22S hardware, but both products use the STM32WLE5 so it shouldn't be too much of a stretch.
